lonestar/analytics/distributed/tc/tc.cpp

```diff
@@ -61,36 +72,74 @@ template <bool async>
 struct TC {
   Graph* graph;
   using DGAccumulatorTy = galois::DGAccumulator<uint64_t>;
-  DGAccumulatorTy& num_triangles;
+  DGAccumulatorTy& numTriangles;
 
-  TC(Graph* _graph, DGAccumulatorTy& _num_triangles)
-      : graph(_graph), num_triangles(_num_triangles) {}
+  TC(Graph* _graph, DGAccumulatorTy& _numTriangles)
+      : graph(_graph), numTriangles(_numTriangles) {}
 
   // use the below line once CPU code is added
-  // void static go(Graph& _graph) {
-  void static go() {
+  void static go(Graph& _graph) {
     unsigned _num_iterations = 0;
-    DGAccumulatorTy num_triangles;
+    DGAccumulatorTy numTriangles;
     syncSubstrate->set_num_round(_num_iterations);
-    num_triangles.reset();
+    numTriangles.reset();
+    // const auto& allNodes = _graph.allNodesWithEdgesRange();
+    const auto& allNodes = _graph.masterNodesRange();
 
+    if (personality == GPU_CUDA) { ///< GPU TC.
 #ifdef GALOIS_ENABLE_GPU
-    if (personality == GPU_CUDA) {
       std::string impl_str(syncSubstrate->get_run_identifier("TC"));
       galois::StatTimer StatTimer_cuda(impl_str.c_str(), regionname);
       StatTimer_cuda.start();
       uint64_t num_local_triangles = 0;
       TC_masterNodes_cuda(num_local_triangles, cuda_ctx);
-      num_triangles += num_local_triangles;
+      numTriangles += num_local_triangles;
       StatTimer_cuda.stop();
-    }
+#else
+      abort();
 #endif
+    } else { ///< CPU TC.
+      galois::do_all(
+          galois::iterate(allNodes), TC(&_graph, numTriangles), galois::steal(),
+          galois::loopname(syncSubstrate->get_run_identifier("TC").c_str()));
+    }
 
-    uint64_t total_triangles = num_triangles.reduce();
+    uint64_t total_triangles = numTriangles.reduce();
     if (galois::runtime::getSystemNetworkInterface().ID == 0) {
       galois::gPrint("Total number of triangles ", total_triangles, "\n");
     }
   }
+
+  void operator()(GNode p1) const {
+    size_t numTriangles_local = 0;
+    for (auto it_p1 : graph->edges(p1)) {
+      GNode p2 = graph->getEdgeDst(it_p1);
+
+      Graph::edge_iterator p1Begin = graph->edge_begin(p1);
+      Graph::edge_iterator p1End   = graph->edge_end(p1);
+      Graph::edge_iterator p2Begin = graph->edge_begin(p2);
+      Graph::edge_iterator p2End   = graph->edge_end(p2);
+      Graph::edge_iterator p1p     = p1Begin;
+      Graph::edge_iterator p2p     = p2Begin;
+      uint32_t p1Dest, p2Dest;
+
+      while (p1p < p1End && p2p < p2End) {
+        p1Dest           = graph->getEdgeDst(p1p);
+        p2Dest           = graph->getEdgeDst(p2p);
+        int32_t nodeDiff = p1Dest - p2Dest;
```

Comment:
I don't particularly like using signed ints to detect termination: problem is that overflow is a concern once the destinations become big enough (e.g. what happens if p1Dest is 0 and p2Dest is 2^31 + 1? you would get an overflow issue). I would change the logic to not have to use arithmetic on the destinations instead if at all possible.

Reply:
Agree. But for now, let's be consistent with what GPU and lonestar cpu is doing.

Reply:
Lonestar CPU is doing it differently actually: I made a comment below about checking the ordered count algorithm.

Reply:
I agree with both of you. Any option works for me. I just tried to make it consistent with GPU codes. So what is your conclusion? If I use unsigned, then GPU codes should be updated, too.

Reply:
GPU has to do things differently due to the threat of warp divergence: CPU and GPU don't have to be consistent.

I would make it consistent with ordered count in Lonestar CPU.

Reply:
I see. Thank you.

Reply:
Agreed